ddtrace/tracer/sampler.go

```diff
@@ -147,3 +154,242 @@ func (ps *prioritySampler) apply(spn *span) {
+		}{}
+		err := json.Unmarshal([]byte(rulesFromEnv), &jsonRules)
+		if err != nil {
+			log.Warn("error parsing DD_TRACE_SAMPLING_RULES: %v", err)
+			return nil
+		}
+		for _, v := range jsonRules {
+			if v.Rate == "" {
+				log.Warn("error parsing rule: rate not provided")
+				continue
+			}
+			rate, err := v.Rate.Float64()
+			if err != nil {
+				log.Warn("error parsing rule: invalid rate: %v", err)
+				continue
+			}
+			switch {
+			case v.Service != "" && v.Operation != "":
+				rules = append(rules, ServiceOperationRule(v.Service, v.Operation, rate))
+			case v.Service != "":
+				rules = append(rules, ServiceRule(v.Service, rate))
+			case v.Operation != "":
+				rules = append(rules, OperationRule(v.Operation, rate))
+			}
+		}
+	}
+	validRules := make([]SamplingRule, 0, len(rules))
+	for _, v := range rules {
+		if !(v.Rate >= 0.0 && v.Rate <= 1.0) {
+			log.Warn("ignoring rule %+v: rate is out of range", v)
+			continue
+		}
+		validRules = append(validRules, v)
+	}
+	return validRules
+}
+
+func sampleRate() float64 {
+	const defaultRate = 0.0
+	v := os.Getenv("DD_TRACE_SAMPLE_RATE")
+	if v == "" {
+		return defaultRate
+	}
+	r, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("using default rate %f because DD_TRACE_SAMPLE_RATE is invalid: %v", defaultRate, err)
+		return defaultRate
+	}
+	if r >= 0.0 && r <= 1.0 {
+		return r
+	}
+	log.Warn("using default rate %f because provided value is out of range: %f", defaultRate, r)
+	return defaultRate
+}
+
+func newRateLimiter(r float64) *rate.Limiter {
+	v := os.Getenv("DD_TRACE_RATE_LIMIT")
+	if v == "" {
+		return nil
+	}
+	l, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("using default rate limit because DD_TRACE_RATE_LIMIT is invalid: %v", err)
+		return nil
+	}
+	switch {
+	case l < 0.0:
+		return nil
+	case l == 0.0:
+		return rate.NewLimiter(0.0, 0)
+	case (r * l) < 1.0:
+		return rate.NewLimiter(rate.Limit(l), 1)
+	default:
+		return rate.NewLimiter(rate.Limit(l), int(math.Ceil(r*l)))
+	}
+}
+
+func (rs *rulesSampler) apply(span *span) bool {
+	var matched bool
+	rate := 0.0
+	for _, rule := range rs.rules {
+		if rule.match(span) {
+			matched = true
+			rate = rule.Rate
+			break
+		}
+	}
+	if !matched {
+		if rs.rate == 0.0 {
+			// no matching rule or global rate, so we want to fall back
+			// to priority sampling
+			return false
+		}
+		rate = rs.rate
+	}
+	// rate sample
+	span.SetTag("_dd.rule_psr", rate)
+	if !sampledByRate(span.TraceID, rate) {
+		span.SetTag(ext.SamplingPriority, ext.PriorityAutoReject)
+		return true
+	}
+	// global rate limit and effective rate calculations
+	defer rs.mu.Unlock()
+	rs.mu.Lock()
+	if ts := time.Now().Unix(); ts > rs.ts {
+		// update "previous rate" and reset
+		if ts-rs.ts == 1 && rs.total > 0 && rs.allowed > 0 {
+			rs.previousRate = float64(rs.allowed) / float64(rs.total)
+		} else {
+			rs.previousRate = 0.0
+		}
+		rs.ts = ts
+		rs.allowed = 0
+		rs.total = 0
+	}
```

Comment:
I feel this needs it's own function and associated tests. Also, can we not use `time.Time` instead of `int64`? It'd be much easier to read.